File: algorithms/medium/incremental_memory_leak.ex
defmodule Solution do
  @spec mem_leak(memory1 :: integer, memory2 :: integer) :: [integer]
  def mem_leak(memory1, memory2) do
    mem_leak(1, memory1, memory2)
  end

  def mem_leak(i, memory1, memory2) do
    cond do
      i > memory1 and i > memory2 -> [i, memory1, memory2]
      memory1 >= memory2 -> mem_leak(i + 1, memory1 - i, memory2)
      true -> mem_leak(i + 1, memory1, memory2 - i)
    end
  end
end

File: algorithms/medium/incremental_memory_leak.rs
impl Solution {
    pub fn mem_leak(mut memory1: i32, mut memory2: i32) -> Vec<i32> {
        let mut i = 1;
        loop {
            if i > memory1 && i > memory2 {
                break;
            }
            if memory1 >= memory2 {
                memory1 -= i;
            } else {
                memory2 -= i;
            }
            i += 1;
        }
        vec![i, memory1, memory2]
    }
}
